ddtrace/internal/_rand.pyx

```diff
@@ -18,38 +18,83 @@ That's 10,000,000 seconds until we hit the period. That's 115 days of
-rand64bits                         167.5956 (1.0)      211.3155 (1.0)      190.2803 (1.0)       9.5815 (1.0)      187.7187 (1.0)      11.4513 (1.0)          15;1        5.2554 (1.0)          52      100000
-random.randbits                    222.7103 (1.33)     367.4459 (1.74)     250.2699 (1.32)     26.5930 (2.78)     242.1607 (1.29)     26.4550 (2.31)          6;1        3.9957 (0.76)         36      100000
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+Name (time in ns)                     Min                 Max                Mean             StdDev              Median                IQR            Outliers  OPS (Mops/s)            Rounds  Iterations
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+rand64bits                        57.5356 (1.0)      151.9838 (1.0)       79.8573 (1.0)      20.5218 (1.0)       73.8342 (1.0)      24.0444 (1.0)         24;10       12.5223 (1.0)         147      100000
+random.getrandbits               111.7341 (1.94)     224.2096 (1.48)     137.4291 (1.72)     24.7072 (1.20)     126.5603 (1.71)     24.1269 (1.00)         15;6        7.2765 (0.58)         82      100000
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 """
 from libc.stdint cimport uint64_t
+import os
 
+from ddtrace.vendor.wrapt import wrap_function_wrapper
 from ddtrace import compat
 
 
-cdef uint64_t x = compat.getrandbits(64) ^ 4101842887655102017
+cdef uint64_t state
+
+
+cpdef _getstate():
+    return state
+
+
+cpdef seed():
+    global state
+    state = compat.getrandbits(64) ^ 4101842887655102017
+
+
+cpdef rand64bits():
+    global state
+    state ^= state >> 21
+    state ^= state << 35
+    state ^= state >> 4
+    return <uint64_t>(state * <uint64_t>2685821657736338717)
+
+
+# Should be available in Python 3.7+
+if hasattr(os, "register_at_fork"):
+    os.register_at_fork(after_in_child=seed)
+
+
+def patch_stdlib_seed():
+    """Patches random.seed() to also reseed our RNG.
+
+    This is done because many libraries when they fork will call random.seed()
+    to reseed the generator for the new process.
+    """
+
+    def patched_seed(func, instance, args, kwargs):
+        seed()
+        return func(*args, **kwargs)
+
+
+    wrap_function_wrapper("random", "seed", patched_seed)
+
 
+seed()
 
-def rand64bits():
-    global x
-    x ^= x >> 21
-    x ^= x << 35
-    x ^= x >> 4
-    return x * <uint64_t>2685821657736338717
+patch_stdlib_seed()
```

Comment:
Yes. It is possible that someone who forks does not call the Python fork hooks. In which case we may still be able to get reseeded if that someone then reseeds the Python PRNG.

Re-seeding multiple times shouldn't be too big an issue since it'll only happen at most once when the process is forked. `uwsgi` actually [reseeds the Python PRNG twice](https://github.com/unbit/uwsgi/blob/3d17bf5115a300d722776a78bf0ef01764556cba/plugins/python/python_plugin.c#L436-L444).

Reply:
got it, ok, 👍